Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp

```diff
@@ -45,6 +52,173 @@ namespace WebKit {
+            auto path = writeIconToTemporaryFile(icon);
+            if (path.isNull()) {
+                m_iconCache.remove(addResult.iterator);
+                return nullptr;
+            }
+            addResult.iterator->value = { 1, WTFMove(path) };
+        } else
+            addResult.iterator->value.first++;
+
+        return std::get<CString>(addResult.iterator->value.second).data();
+    }
+
+    GBytes* iconBytes(const String& iconURL, const RefPtr<WebCore::Image>& icon)
+    {
+        if (!icon)
+            return nullptr;
+
+        auto writeIconToBuffer = [](const RefPtr<WebCore::Image>& icon) -> GRefPtr<GBytes> {
+            auto nativeImage = icon->nativeImage();
+            if (!nativeImage)
+                return nullptr;
+
+            auto surface = nativeImage->platformImage();
+            if (!surface)
+                return nullptr;
+
+            GRefPtr<GByteArray> buffer = adoptGRef(g_byte_array_new());
+            auto status = cairo_surface_write_to_png_stream(surface.get(), [](void* userData, const unsigned char* data, unsigned length) -> cairo_status_t {
+                auto* buffer = static_cast<GByteArray*>(userData);
+                g_byte_array_append(buffer, data, length);
+                return CAIRO_STATUS_SUCCESS;
+            }, buffer.get());
+
+            if (status != CAIRO_STATUS_SUCCESS)
+                return nullptr;
+
+            return adoptGRef(g_byte_array_free_to_bytes(buffer.leakRef()));
+        };
+
+        auto addResult = m_iconCache.add(iconURL, std::pair<uint32_t, GRefPtr<GBytes>>({ 0, nullptr }));
+        if (addResult.isNewEntry) {
+            auto bytes = writeIconToBuffer(icon);
+            if (!bytes) {
+                m_iconCache.remove(addResult.iterator);
+                return nullptr;
+            }
+            addResult.iterator->value = { 1, WTFMove(bytes) };
+        } else
+            addResult.iterator->value.first++;
+
+        return std::get<GRefPtr<GBytes>>(addResult.iterator->value.second).get();
+    }
+
+    void unuseIcon(const String& iconURL)
+    {
+        auto it = m_iconCache.find(iconURL);
+        if (it == m_iconCache.end())
+            return;
+
+        if (!it->value.first)
+            return;
+
+        bool isNull = WTF::switchOn(it->value.second,
+            [](const CString& path) {
+                return path.isNull();
+            },
+            [](const GRefPtr<GBytes>& bytes) {
+                return !bytes.get();
+            });
+        if (isNull)
+            return;
+
+        if (--it->value.first)
+            return;
+
+        m_timer.startOneShot(5_min);
+    }
+
+    void removeUnusedIcons(bool force)
+    {
+        m_iconCache.removeIf([force](auto& it) -> bool {
+            if (!it.value.first || force) {
+                WTF::switchOn(it.value.second,
+                    [](const CString& path) {
+                        if (!path.isNull())
+                            unlink(path.data());
```

Comment:
I like to check for errors whenever possible, since it makes debugging so much easier when things go unexpectedly wrong:

```suggestion
                        if (!path.isNull())
                            if (unlink(path.data()) == -1)
                                WTFLogAlways("Failed to remove cached icon %s: %s", path.data(), safeStrerror(errno).data());
```

Reply:
This is very unlikely to fail, but ok.